test/e2e/ipam/ipamhelper.go

```diff
@@ -0,0 +1,225 @@
+		Byf("Setting clusterctl variable %s to %s", variable, ip)
+		variables[variable] = ip
+	}
+
+	// Create a new clusterctl config file based on the passed file and add the new variables for the IPs.
+	modifiedClusterctlConfigPath := fmt.Sprintf("%s-%s.yaml", strings.TrimSuffix(clusterctlConfigPath, ".yaml"), rand.String(16))
+	Byf("Writing a new clusterctl config to %s", modifiedClusterctlConfigPath)
+	copyAndAmendClusterctlConfig(ctx, copyAndAmendClusterctlConfigInput{
+		ClusterctlConfigPath: clusterctlConfigPath,
+		OutputPath:           modifiedClusterctlConfigPath,
+		Variables:            variables,
+	})
+
+	return modifiedClusterctlConfigPath, ipamService.cleanup
+}
+
+type ipamService struct {
+	client          client.Client
+	ipAddressClaims []ipamv1.IPAddressClaim
+	skipCleanup     bool
+}
+
+func newIPAMService(e2eIPAMKubeconfig string, skipCleanup bool) *ipamService {
+	kubeConfig, err := os.ReadFile(filepath.Clean(e2eIPAMKubeconfig))
+	Expect(err).ToNot(HaveOccurred())
+
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(kubeConfig)
+	Expect(err).ToNot(HaveOccurred())
+
+	ipamClient, err := client.New(restConfig, client.Options{Scheme: ipamScheme})
+	Expect(err).ToNot(HaveOccurred())
+
+	return &ipamService{
+		ipAddressClaims: []ipamv1.IPAddressClaim{},
+		client:          ipamClient,
+		skipCleanup:     skipCleanup,
+	}
+}
+
+func (s *ipamService) claimIPAddress(ctx context.Context) string {
+	claim := &ipamv1.IPAddressClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "ipclaim-" + rand.String(32),
+			Namespace: metav1.NamespaceDefault,
+			Annotations: map[string]string{
+				"prow.k8s.io/build-id": os.Getenv("BUILD_ID"),
+				"prow.k8s.io/job":      os.Getenv("JOB_NAME"),
+			},
+		},
+		Spec: ipamv1.IPAddressClaimSpec{
+			PoolRef: corev1.TypedLocalObjectReference{
+				APIGroup: ptr.To("ipam.cluster.x-k8s.io"),
+				Kind:     "InClusterIPPool",
+				Name:     "capv-e2e-ippool",
+			},
+		},
+	}
+
+	// Create an IPAddressClaim
+	Byf("Creating IPAddressClaim %s/%s", claim.GetNamespace(), claim.GetName())
+	Expect(s.client.Create(ctx, claim)).To(Succeed())
+
+	// Wait for the IPAddressClaim to refer an IPAddress.
+	Eventually(func() error {
+		if err := s.client.Get(ctx, client.ObjectKeyFromObject(claim), claim); err != nil {
+			return err
+		}
+
+		if claim.Status.AddressRef.Name == "" {
+			return fmt.Errorf("IPAddressClaim does not yet have an IP")
+		}
+
+		return nil
+	}, time.Second*30, time.Second).Should(Succeed())
+
+	ip := &ipamv1.IPAddress{}
+
+	Expect(s.client.Get(ctx, client.ObjectKey{Namespace: claim.GetNamespace(), Name: claim.Status.AddressRef.Name}, ip)).To(Succeed())
+	Expect(ip.Spec.Address).ToNot(Equal(""))
+
+	// Store claim inside the service so the cleanup function knows what to delete.
+	s.ipAddressClaims = append(s.ipAddressClaims, *claim)
+
+	return ip.Spec.Address
+}
+
+// cleanup deletes the IPAddressClaims created by the service if the test succeeded.
+func (s *ipamService) cleanup() {
```

Comment:
If we do this let's make it very clear please, e.g. via an additional flag. Otherwise it can be a very magic side effect that other (and potentially independent) IPAddressClaims in our IPAM cluster are deleted by running e2e tests.

Wondering if it should be a separate binary altogether (that we can also run in a separate ProwJob). We have a bunch of other objects to cleanup (like VMs)

Reply:
Yep, the general cleanup is also required for other projects we share the code + env with (In summary: CAPV, CPI, image-builder) and these require different credentials too.

We currently have to cleanup this manually but this should maybe be a different periodic or a cronjob instead.